### PR TITLE
Fixes case sensitivity on Startup names

### PR DESCRIPTION
**Pull request by jkotalik**
Fix for #877 
I now check when there is an environmentName that
1. The string starts with methodNameStart
2. The string has the environmentName right after methodNameStart
3. The string has the methodNameEnd right after environmentName 

Also we should add StringSegments?

### CODE REVIEW

File: src/Microsoft.AspNetCore.Hosting/Internal/StartupLoader.cs [hunk 1]
```diff
@@ -65,7 +69,7 @@ public static StartupMethods LoadMethods(IServiceProvider hostingServiceProvider
 
         public static Type FindStartupType(string startupAssemblyName, string environmentName)
         {
-            if (string.IsNullOrEmpty(startupAssemblyName))
+           if (string.IsNullOrEmpty(startupAssemblyName))
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Hosting/Internal/StartupLoader.cs]: formatting?

File: src/Microsoft.AspNetCore.Hosting/Internal/StartupLoader.cs [hunk 2]
```diff
@@ -117,51 +121,71 @@ public static Type FindStartupType(string startupAssemblyName, string environmen
...
             var methods = startupType.GetMethods(BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static);
-            var selectedMethods = methods.Where(method => method.Name.Equals(methodNameWithEnv)).ToList();
-            if (selectedMethods.Count > 1)
+
+            if (environmentName != null)
             {
-                throw new InvalidOperationException(string.Format("Having multiple overloads of method '{0}' is not supported.", methodNameWithEnv));
+                var selectedMethods = methods.Where(method =>
+                {
+                    if (!method.Name.StartsWith(methodNameStart, StringComparison.Ordinal))
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Hosting/Internal/StartupLoader.cs]: Why case sensitive for any part of the method name? It's weird to just do it for one part.

File: src/Microsoft.AspNetCore.Hosting/Internal/StartupLoader.cs [hunk 3]
```diff
@@ -117,51 +121,71 @@ public static Type FindStartupType(string startupAssemblyName, string environmen
...
+                }
+                methodInfo = selectedMethods.FirstOrDefault();
             }
-            if (selectedMethods.Count == 0)
+
+            if (methodInfo == null)
             {
-                selectedMethods = methods.Where(method => method.Name.Equals(methodNameWithNoEnv)).ToList();
+                var methodNameWithNoEnv = methodNameStart + methodNameEnd;
+                var selectedMethods = methods.Where(method => method.Name.Equals(methodNameWithNoEnv)).ToList();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Hosting/Internal/StartupLoader.cs]: specify compairison

File: src/Microsoft.AspNetCore.Hosting/Internal/StartupLoader.cs (hunk 2 above)
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Hosting/Internal/StartupLoader.cs]: I thought we didn't want to check against Configure, Services, and Container for case sensitivity. Hence the check on only the env name.

File: test/Microsoft.AspNetCore.Hosting.Tests/StartupManagerTests.cs [hunk 4]
```diff
@@ -88,6 +88,29 @@ public void StartupClassAddsConfigureServicesToApplicationServices(string enviro
             Assert.Equal(environment, options.Environment);
         }
 
+        [Theory]
+        [InlineData("caseinsensitive")]
+        [InlineData("CaseInsensitive")]
+        [InlineData("CASEINSENSITIVE")]
+        [InlineData("CaSEiNSENsitiVE")]
+        public void StartupClassAddsConfigureServicesToApplicationServicesCaseInsensitive(string environment)
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.Hosting.Tests/StartupManagerTests.cs]: Does this test configure as well?

**Comment by jkotalik** [comment]
I've realized now that there are two concerns here, both in class and method finding. I'll update it to handle class finding.

**Comment by jkotalik** [comment]
@davidfowl Should this change have case insensitivity for only the Environment Name, or the whole method/class name? If we did the latter, we would break people that had method names with different cases. ex startupEnvName and StartupEnvName would both be found.

**Comment by jkotalik** [comment]
@Tratcher too

**Pull request merged**